src/sas/sascalc/pr/calc.py

```diff
@@ -0,0 +1,368 @@
+
+    return total/nslice_d * d_max
+
+@njit('f8(f8[:], f8, u8)')
+def int_pr_square(pars, d_max, nslice):
+    """
+    Regularization term calculated from the expansion.
+
+    :param pars: c-parameters.
+    :param d_max: d_max.
+    :param nslice: nslice.
+
+    :return:  Regularization term calculated from the expansion.
+    """
+    nslice_d = np.float64(nslice)
+    i_r = np.arange(nslice)
+    r = d_max/nslice_d * i_r
+    values = pr(pars, d_max, r)
+    total = np.sum(values ** 2)
+
+    return total/nslice_d * d_max
+
+@njit('f8(f8[:], f8, u8)')
+def int_pr(pars, d_max, nslice):
+    """
+    Integral of P(r).
+
+    :param pars: c-parameters.
+    :param d_max: d_max.
+    :param nslice: nslice.
+
+    :return: Integral of P(r).
+    """
+    nslice_d = np.float64(nslice)
+
+    i_r = np.arange(nslice)
+    r = d_max/nslice_d * i_r
+    values = pr(pars, d_max, r)
+
+    total = np.sum(values)
+
+    return total/nslice_d * d_max
+
+@njit('u8(f8[:], f8, u8)')
+def npeaks(pars, d_max, nslice):
+    """
+    Get the number of P(r) peaks.
+
+    :param pars: c-parameters.
+    :param d_max: d_max.
+    :param nslice: nslice.
+
+    :return: Number of P(r) peaks.
+    """
+    i_r = np.arange(nslice)
+    r = d_max/np.float64(nslice) * i_r
+    values = pr(pars, d_max, r)
+
+    # Build an index vector with True for ascending and false for flat or descending
+    pos = values[:-1] < values[1:]
+    # Count the number of slope changes
+    count = np.sum((pos[:-1] != pos[1:]) & pos[:-1])
+    # Check if starting descending or ending ascending
+    count += 1 - pos[0] + pos[-1]
+
+    return count
+
+
+
+
+@njit('f8(f8[:], f8, u8)')
+def positive_integral(pars, d_max, nslice):
+    """
+    Get the fraction of the integral of P(r) over the whole
+    range of r that is above 0.
+    A valid P(r) is defined as being positive for all r.
+
+    :param pars: c-parameters.
+    :param d_max: d_max.
+    :param nslice: nslice.
+
+    :return: The fraction of the integral of P(r) over the whole
+    range of r that is above 0.
+    """
+    nslice_d = np.float64(nslice)
+    i_r = np.arange(nslice)
+    r = d_max/nslice_d * i_r
+    values = pr(pars, d_max, r)
+
+    total = np.sum(np.fabs(values))
+    total_pos = np.sum(values[values > 0.0])
+
+    return total_pos / total
+
+@njit('f8(f8[:], f8[:,:], f8, u8)')
+def positive_errors(pars, err, d_max, nslice):
+    """
+    Get the fraction of the integral of P(r) over the whole range
+    of r that is at least one sigma above 0.
+
+    :param pars: c-parameters.
+    :param err: error terms.
+    :param d_max: d_max.
+    :param nslice: nslice.
+
+    :return: The fraction of the integral of P(r) over the whole range
+    of r that is at least one sigma above 0.
+    """
+    nslice_d = np.float64(nslice)
+    i_r = np.arange(nslice)
+    r = d_max/nslice_d * i_r
+    pr_vals = pr_err(pars, err, d_max, r)
+
+    pr_val = pr_vals[0, :]
+    pr_val_err = pr_vals[1, :]
+
```

Comment:
Can use `pr_val, pr_val_err = pr_err(pars, err, d_max, r)` even if `pr_err` returns a matrix rather than a tuple, but the above form won't work for a tuple.